src/main/java/net/neoforged/neoforge/flag/Flag.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.flag;
+
+import com.google.common.collect.Maps;
+import com.mojang.serialization.Codec;
+import io.netty.buffer.ByteBuf;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.resources.ResourceLocation;
+
+/**
+ * Class describing a specific flag.
+ */
+public final class Flag {
+    public static final Codec<Flag> CODEC = ResourceLocation.CODEC.xmap(Flag::of, flag -> flag.identifier);
+    public static final StreamCodec<ByteBuf, Flag> STREAM_CODEC = ResourceLocation.STREAM_CODEC.map(Flag::of, flag -> flag.identifier);
+
+    private static final Map<ResourceLocation, Flag> FLAGS = Maps.newConcurrentMap();
+    private static final Collection<Flag> FLAGS_VIEW = Collections.unmodifiableCollection(FLAGS.values());
+
+    private final ResourceLocation identifier;
+
+    private Flag(ResourceLocation identifier) {
+        this.identifier = identifier;
+    }
+
+    /**
+     * @return The owning mod id/namespace for this {@link Flag flag}.
+     */
+    public String namespace() {
+        return identifier.getNamespace();
+    }
+
+    /**
+     * @return The unique identifier for this {@link Flag flag}.
+     * @apiNote This identifier does not need to be globally unique, only unique to the owning mod.
+     */
+    public String identifier() {
+        return identifier.getPath();
+    }
+
+    public String toStringShort() {
+        return identifier.toString();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj)
+            return true;
+        if (obj.getClass() == Flag.class)
```

Comment:
`obj instanceof Flag`???

Reply:
`Flag` is final, there is no benefit to paying the overhead of an instanceof

Reply:
> `Flag` is final, there is no benefit to paying the overhead of an instanceof

Overhead? Could you elaborate?

Reply:
`instanceof` is more efficient. I would recommend benchmarking before premature optimization.

Reply:
I was very certain that `instanceof` had a larger overhead than a class literal comparison. But whatever, I'm probably just being pedantic.

Reply:
So just to be clear, we are in favor of reverting this back to being an `instanceof`?